v2/bigtable-cdc-to-hbase/src/main/java/com/google/cloud/teleport/v2/templates/transforms/ConvertChangeStream.java

```diff
@@ -0,0 +1,252 @@
+
+    private String hbaseQualifier;
+    private String cbtQualifier;
+    private boolean twoWayReplication;
+
+    public ConvertChangeStreamMutationFn(
+        boolean twoWayReplicationFlag, String cbtQualifierInput, String hbaseQualifierInput) {
+      twoWayReplication = twoWayReplicationFlag;
+      hbaseQualifier = hbaseQualifierInput;
+      cbtQualifier = cbtQualifierInput;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+
+      ChangeStreamMutation mutation = c.element().getValue();
+
+      // Skip element if it was replicated from HBase.
+      if (twoWayReplication && isHbaseReplicated(mutation, hbaseQualifier)) {
+        return;
+      }
+      RowMutations hbaseMutations = convertToRowMutations(mutation);
+
+      // Append origin information to mutations.
+      if (twoWayReplication) {
+        appendOriginInfoToMutations(hbaseMutations, cbtQualifier);
+      }
+
+      c.output(KV.of(hbaseMutations.getRow(), hbaseMutations));
+    }
+
+    /**
+     * Checks if mutation was replicated from HBase.
+     *
+     * @param mutation from change stream
+     * @return true if mutation was replicated from hbase
+     */
+    private boolean isHbaseReplicated(ChangeStreamMutation mutation, String hbaseQualifierInput) {
+      List<Entry> mutationEntries = mutation.getEntries();
+
+      if (mutationEntries.size() == 0) {
+        return false;
+      }
+      Entry lastEntry = mutationEntries.get(mutationEntries.size() - 1);
+
+      if (lastEntry instanceof DeleteCells) {
+        if (((DeleteCells) lastEntry)
+            .getQualifier()
+            .equals(ByteString.copyFromUtf8(hbaseQualifierInput))) {
+          LOG.info(
+              "Filtering: {}",
+              mutation.toRowMutation("").toProto(RequestContext.create("", "", "")));
+          Metrics.counter("HbaseRepl", "mutations_filtered_from_hbase").inc();
+          return true;
+        }
+      }
+      Metrics.counter("HbaseRepl", "mutations_valid_from_bigtable").inc();
+      return false;
+    }
+
+    static byte[] convertUtf8String(String utf8) {
+      return utf8.getBytes(StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Appends origin information to row mutation.
+     *
+     * @param hbaseMutations row mutation to append origin info to
+     * @param cbtQualifierInput origin info string denoting mutation is from bigtable
+     * @throws IOException
+     */
+    private void appendOriginInfoToMutations(RowMutations hbaseMutations, String cbtQualifierInput)
+        throws IOException {
+      // Get last mutation to deduce last column family from
+      Mutation lastMutation =
+          hbaseMutations.getMutations().get(hbaseMutations.getMutations().size() - 1);
+      List<String> familyList = (List<String>) lastMutation.getFingerprint().get("families");
+      String lastEntryCf = familyList.get(familyList.size() - 1);
```

Comment:
nit: pulling the code for identify the lastEntryCf into a helper might help with readability here

Reply:
Done